File: tools/xlsxread/src/transform.rs
/// Row transformation: ascii normalisation, score regex parsing, validation.
///
/// `to_ascii` replicates build-lib.js `toAscii` exactly:
///   str.normalize("NFD").replace(/[̀-ͯ]/g,"").replace(/đ/gi,"d").toLowerCase()
use std::collections::HashMap;

use regex::Regex;
use unicode_normalization::UnicodeNormalization;

use crate::config::{DatasetConfig, ValidationCfg};
use crate::error::BuildError;

// ---------------------------------------------------------------------------
// Compiled score patterns (built once at startup from config)
// ---------------------------------------------------------------------------

pub struct CompiledPatterns {
    /// Ordered list so INSERT column order is deterministic
    pub patterns: Vec<(String, Regex)>,
}

impl CompiledPatterns {
    pub fn new(scores: &HashMap<String, String>) -> Result<Self, BuildError> {
        let mut patterns = Vec::with_capacity(scores.len());
        for (field, src) in scores {
            let re = Regex::new(src).map_err(|e| BuildError::Regex {
                pattern: src.clone(),
                source: e,
            })?;
            patterns.push((field.clone(), re));
        }
        // Sort for deterministic order across HashMap iteration
        patterns.sort_by(|a, b| a.0.cmp(&b.0));
        Ok(Self { patterns })
    }
}

// ---------------------------------------------------------------------------
// to_ascii — must be byte-for-byte equivalent to build-lib.js toAscii
// ---------------------------------------------------------------------------

/// Normalise a Vietnamese name to an ASCII slug.
///
/// Algorithm mirrors the JavaScript `toAscii` in build-lib.js:
///   1. NFD decompose (splits base + combining diacritics)
///   2. Drop all Unicode combining marks (U+0300–U+036F)
///   3. Replace đ/Đ with d (NFD does not decompose đ)
///   4. Lowercase
pub fn to_ascii(s: &str) -> String {
    // Step 1 + 2: NFD then filter out combining marks (Unicode category M)
    let decomposed: String = s
        .nfd()
        .filter(|c| !('\u{0300}'..='\u{036f}').contains(c))
        .collect();

    // Step 3: đ/Đ are not decomposed by NFD — replace explicitly
    let replaced = decomposed.replace(['đ', 'Đ'], "d");

    // Step 4: lowercase
    replaced.to_lowercase()
}

// ---------------------------------------------------------------------------
// Parsed row ready for DB insert
// ---------------------------------------------------------------------------

pub struct ParsedRow {
    pub so_bao_danh: String,
    pub ho_ten: String,
    pub ho_ten_ascii: String,
    pub ngay_sinh: Option<String>,
    /// Subject field → float value; absent subjects not in map → NULL
    pub scores: HashMap<String, f64>,
}

// ---------------------------------------------------------------------------
// Row validation — mirrors the per-script skip logic
// ---------------------------------------------------------------------------

/// Returns `None` when the row should be skipped entirely (before sourceRows counter).
/// Returns `Some(reason)` when the row should be counted as sourceRows but skipped.
#[derive(Debug, PartialEq, Eq)]
pub enum SkipReason {
    /// Row is fully blank (data-old2 only, before sourceRows counter)
    BlankRow,
    /// soBaoDanh or hoTen empty/missing
    EmptyField,
    /// soBaoDanh contains non-digit characters (data-old / data-old2 guard)
    NonNumericSbd,
}

/// Validates a raw cell slice against the dataset's `ValidationCfg`.
/// Returns `Ok(())` on pass, `Err(SkipReason)` on fail.
pub fn validate_row(
    ho_ten: &str,
    so_bao_danh: &str,
    cfg: &ValidationCfg,
    strip_blank_rows: bool,
    all_blank: bool,
) -> Result<(), SkipReason> {
    // data-old2: skip fully blank rows BEFORE counting sourceRows
    if strip_blank_rows && all_blank {
        return Err(SkipReason::BlankRow);
    }

    if cfg.require_nonempty_sbd && so_bao_danh.is_empty() {
        return Err(SkipReason::EmptyField);
    }
    if cfg.require_nonempty_name && ho_ten.is_empty() {
        return Err(SkipReason::EmptyField);
    }
    if cfg.require_numeric_sbd && !so_bao_danh.chars().all(|c| c.is_ascii_digit()) {
        return Err(SkipReason::NonNumericSbd);
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Score parsing — mirrors build-lib.js parseScores
// ---------------------------------------------------------------------------

/// Parse a DIEM_THI cell string and extract matching subject scores.
pub fn parse_scores(diem_thi: &str, patterns: &CompiledPatterns) -> HashMap<String, f64> {
    let mut out = HashMap::new();
    for (field, re) in &patterns.patterns {
        if let Some(caps) = re.captures(diem_thi) {
            if let Some(m) = caps.get(1) {
                if let Ok(v) = m.as_str().parse::<f64>() {
                    if v.is_finite() {
                        out.insert(field.clone(), v);
                    }
                }
            }
        }
    }
    out
}

// ---------------------------------------------------------------------------
// Full row transform
// ---------------------------------------------------------------------------

/// Extract and transform one spreadsheet row into a `ParsedRow`.
/// `raw` is the full cell slice; column indices come from `cfg.columns`.
pub fn transform_row(
    raw: &[calamine::Data],
    cfg: &DatasetConfig,
    patterns: &CompiledPatterns,
) -> ParsedRow {
    let get = |idx: usize| -> String {
        raw.get(idx)
            .map(|cell| cell.to_string().trim().to_owned())
            .unwrap_or_default()
    };

    let ho_ten = get(cfg.columns.ho_ten);
    let ngay_sinh = get(cfg.columns.ngay_sinh);
    let so_bao_danh = get(cfg.columns.so_bao_danh);
    let diem_thi = raw
        .get(cfg.columns.diem_thi)
        .map(|c| c.to_string())
        .unwrap_or_default();

    let ho_ten_ascii = to_ascii(&ho_ten);
    let scores = parse_scores(&diem_thi, patterns);
    let ngay_sinh_opt = if ngay_sinh.is_empty() {
        None
    } else {
        Some(ngay_sinh)
    };

    ParsedRow {
        so_bao_danh,
        ho_ten,
        ho_ten_ascii,
        ngay_sinh: ngay_sinh_opt,
        scores,
    }
}

// ---------------------------------------------------------------------------
// Unit tests — 20 cases for to_ascii (real Vietnamese names)
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    // Helper: assert to_ascii(input) == expected
    fn check(input: &str, expected: &str) {
        assert_eq!(
            to_ascii(input),
            expected,
            "to_ascii({input:?}) expected {expected:?}"
        );
    }

    #[test]
    fn ascii_plain_latin() {
        check("Nguyen Van A", "nguyen van a");
    }

    #[test]
    fn ascii_nguyen_thi_hoa() {
        check("Nguyễn Thị Hoa", "nguyen thi hoa");
    }

    #[test]
    fn ascii_tran_van_duc() {
        // đ/Đ replacement
        check("Trần Văn Đức", "tran van duc");
    }

    #[test]
    fn ascii_le_thi_my_duyen() {
        check("Lê Thị Mỹ Duyên", "le thi my duyen");
    }

    #[test]
    fn ascii_pham_thi_lan() {
        check("Phạm Thị Lan", "pham thi lan");
    }

    #[test]
    fn ascii_bui_thi_thu() {
        check("Bùi Thị Thu", "bui thi thu");
    }

    #[test]
    fn ascii_hoang_van_truong() {
        check("Hoàng Văn Trường", "hoang van truong");
    }

    #[test]
    fn ascii_do_thi_ngan() {
        // Đ uppercase at start
        check("Đỗ Thị Ngân", "do thi ngan");
    }

    #[test]
    fn ascii_nguyen_van_khanh() {
        check("Nguyễn Văn Khánh", "nguyen van khanh");
    }

    #[test]
    fn ascii_trinh_thi_bich_ngoc() {
        check("Trịnh Thị Bích Ngọc", "trinh thi bich ngoc");
    }

    #[test]
    fn ascii_vu_thi_dieu() {
        // ề = e + combining grave + combining circumflex (after NFD)
        check("Vũ Thị Diệu", "vu thi dieu");
    }

    #[test]
    fn ascii_nguyen_thi_tuong_vi() {
        check("Nguyễn Thị Tường Vi", "nguyen thi tuong vi");
    }

    #[test]
    fn ascii_lowercase_d_stroke() {
        // Lowercase đ → d
        check("đặng thị hằng", "dang thi hang");
    }

    #[test]
    fn ascii_uppercase_d_stroke() {
        check("ĐẶNG THỊ HẰNG", "dang thi hang");
    }

    #[test]
    fn ascii_mixed_case() {
        check("NGUYỄN VĂN AN", "nguyen van an");
    }

    #[test]
    fn ascii_tran_thi_kim_anh() {
        check("Trần Thị Kim Anh", "tran thi kim anh");
    }

    #[test]
    fn ascii_nguyen_thi_phuong_thao() {
        check("Nguyễn Thị Phương Thảo", "nguyen thi phuong thao");
    }

    #[test]
    fn ascii_le_van_long() {
        check("Lê Văn Long", "le van long");
    }

    #[test]
    fn ascii_vo_thi_xuan_mai() {
        check("Võ Thị Xuân Mai", "vo thi xuan mai");
    }

    #[test]
    fn ascii_empty_string() {
        check("", "");
    }

    // --- Score parsing tests ---

    fn make_patterns() -> CompiledPatterns {
        let mut map = HashMap::new();
        map.insert("toan".into(), r"Toán:\s*(\d+(?:\.\d+)?)".into());
        map.insert("ngu_van".into(), r"Ngữ văn:\s*(\d+(?:\.\d+)?)".into());
        map.insert("vat_ly".into(), r"Vật lí:\s*(\d+(?:\.\d+)?)".into());
        CompiledPatterns::new(&map).unwrap()
    }

    #[test]
    fn parse_scores_single() {
        let p = make_patterns();
        let s = "Toán: 8.5";
        let scores = parse_scores(s, &p);
        assert_eq!(scores.get("toan"), Some(&8.5));
        assert!(scores.get("ngu_van").is_none());
    }

    #[test]
    fn parse_scores_multiple() {
        let p = make_patterns();
        let s = "Toán: 7.25  Ngữ văn: 6.0  Vật lí: 9";
        let scores = parse_scores(s, &p);
        assert_eq!(scores.get("toan"), Some(&7.25));
        assert_eq!(scores.get("ngu_van"), Some(&6.0));
        assert_eq!(scores.get("vat_ly"), Some(&9.0));
    }

    #[test]
    fn parse_scores_empty_cell() {
        let p = make_patterns();
        let scores = parse_scores("", &p);
        assert!(scores.is_empty());
    }

    // --- Validation tests ---

    fn default_validation() -> ValidationCfg {
        ValidationCfg {
            require_numeric_sbd: false,
            require_nonempty_name: true,
            require_nonempty_sbd: true,
        }
    }

    #[test]
    fn validate_ok() {
        let v = default_validation();
        assert!(validate_row("Nguyen Van A", "12345678", &v, false, false).is_ok());
    }

    #[test]
    fn validate_empty_sbd() {
        let v = default_validation();
        assert_eq!(
            validate_row("Nguyen Van A", "", &v, false, false),
            Err(SkipReason::EmptyField)
        );
    }

    #[test]
    fn validate_empty_name() {
        let v = default_validation();
        assert_eq!(
            validate_row("", "12345678", &v, false, false),
            Err(SkipReason::EmptyField)
        );
    }

    #[test]
    fn validate_non_numeric_sbd_rejected() {
        let mut v = default_validation();
        v.require_numeric_sbd = true;
        assert_eq!(
            validate_row("Nguyen Van A", "12AB5678", &v, false, false),
            Err(SkipReason::NonNumericSbd)
        );
    }

    #[test]
    fn validate_numeric_sbd_accepted() {
        let mut v = default_validation();
        v.require_numeric_sbd = true;
        assert!(validate_row("Nguyen Van A", "12345678", &v, false, false).is_ok());
    }

    #[test]
    fn validate_blank_row_skipped() {
        let v = default_validation();
        // strip_blank_rows=true AND all_blank=true → BlankRow
        assert_eq!(
            validate_row("", "", &v, true, true),
            Err(SkipReason::BlankRow)
        );
    }
}

File: go-parser/internal/transform/transform.go
// Package transform performs row transformation: ASCII normalisation, score
// regex parsing, and validation — a port of parser/src/transform.rs.
//
// ToAscii replicates build-lib.js toAscii exactly:
//
//	str.normalize("NFD").replace(/[̀-ͯ]/g,"").replace(/đ/gi,"d").toLowerCase()
package transform

import (
	"errors"
	"math"
	"strconv"
	"strings"

	"golang.org/x/text/unicode/norm"

	"github.com/tiennm99/thptqg/go-parser/internal/config"
	"github.com/tiennm99/thptqg/go-parser/internal/reader"
	"github.com/tiennm99/thptqg/go-parser/internal/schema"
)

// ToAscii normalises a Vietnamese name to an ASCII slug.
//
//  1. NFD decompose (splits base + combining diacritics)
//  2. Drop combining marks in U+0300..U+036F
//  3. Replace đ/Đ with d (NFD does not decompose them)
//  4. Lowercase
//
// Step 2 filters a LITERAL CODEPOINT RANGE, not a Unicode category. The inline
// comment at transform.rs:53 says "Unicode category M", but the code at :56 is
// the specification and it checks '\u{0300}'..='\u{036f}'. unicode.Is(unicode.Mn, r)
// is strictly broader and would strip marks Rust keeps, silently changing
// ho_ten_ascii — the column the site's accent-insensitive search runs on.
//
// Step order matters and mirrors transform.rs:54-63: the đ/Đ replacement happens
// before lowercasing.
func ToAscii(s string) string {
	if s == "" {
		return ""
	}
	decomposed := norm.NFD.String(s)

	var b strings.Builder
	b.Grow(len(decomposed))
	for _, r := range decomposed {
		if r >= 0x0300 && r <= 0x036F {
			continue // combining mark, in the range Rust drops
		}
		switch r {
		case 'đ', 'Đ':
			b.WriteByte('d')
		default:
			b.WriteRune(r)
		}
	}
	return strings.ToLower(b.String())
}

// ParsedRow is one row ready for insertion.
type ParsedRow struct {
	SoBaoDanh  string
	HoTen      string
	HoTenAscii string
	NgaySinh   *string
	// TenCumThi is 2016 only: examination cluster name (TEN_CUMTHI column).
	TenCumThi *string
	// GioiTinh is 2016 only: gender, normalised to "Nam"/"Nữ" or nil.
	GioiTinh *string
	// Scores maps subject field -> value. Absent subjects are simply missing and
	// bind NULL.
	Scores map[string]float64
}

// SkipReason says why a row was skipped, or SkipNone when it passed.
//
// The distinction is load-bearing for the printed counters, and the two
// non-blank reasons are counted as source rows while BlankRow is not — but note
// that split lives in the CALLER, not here. parser/src/main.rs has two call
// sites with opposite outcomes for BlankRow: at :135-137 it returns before the
// counter at :140, while at :151 it matches Err(BlankRow) => {} and falls
// through to transform and insert. The build loop must reproduce both.
type SkipReason int

const (
	SkipNone SkipReason = iota
	// SkipBlankRow: row is fully blank (2017-old2 only, checked before the
	// source-row counter).
	SkipBlankRow
	// SkipEmptyField: so_bao_danh or ho_ten empty/missing.
	SkipEmptyField
	// SkipNonNumericSbd: so_bao_danh contains non-digit characters
	// (2017-old / 2017-old2 guard).
	SkipNonNumericSbd
)

func (s SkipReason) String() string {
	switch s {
	case SkipNone:
		return "none"
	case SkipBlankRow:
		return "blank_row"
	case SkipEmptyField:
		return "empty_field"
	case SkipNonNumericSbd:
		return "non_numeric_sbd"
	}
	return "unknown"
}

// ValidateRow checks a row against the dataset's validation rules.
//
// Signature mirrors transform.rs:101-107, taking stripBlankRows and allBlank
// explicitly; a shorter signature could not express both blank-row paths.
func ValidateRow(hoTen, soBaoDanh string, cfg *config.ValidationCfg, stripBlankRows, allBlank bool) SkipReason {
	// 2017-old2: skip fully blank rows BEFORE counting source rows.
	if stripBlankRows && allBlank {
		return SkipBlankRow
	}
	if cfg.RequireNonemptySbd && soBaoDanh == "" {
		return SkipEmptyField
	}
	if cfg.RequireNonemptyName && hoTen == "" {
		return SkipEmptyField
	}
	if cfg.RequireNumericSbd && !allASCIIDigits(soBaoDanh) {
		return SkipNonNumericSbd
	}
	return SkipNone
}

// allASCIIDigits mirrors Rust's chars().all(|c| c.is_ascii_digit()).
//
// Deliberately not strconv.Atoi: Atoi accepts a leading sign, so "+123" would
// pass a check Rust rejects. Empty input returns true, matching Rust's all() on
// an empty iterator — the empty case is caught earlier by RequireNonemptySbd.
func allASCIIDigits(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// ParseScores extracts subject scores from a DIEM_THI cell.
//
// Every one of the 16 patterns runs against every dataset; a subject absent from
// a given exam year never matches and stays NULL. Matching is unanchored
// first-match, like Rust's Regex::captures.
func ParseScores(diemThi string) map[string]float64 {
	out := make(map[string]float64)
	if diemThi == "" {
		return out
	}
	for field, re := range schema.ScorePatterns {
		m := re.FindStringSubmatch(diemThi)
		if m == nil {
			continue
		}
		v, err := strconv.ParseFloat(m[1], 64)
		if err != nil {
			continue
		}
		// Unreachable given the pattern shape, but kept for parity with
		// transform.rs:136 (is_finite).
		if math.IsInf(v, 0) || math.IsNaN(v) {
			continue
		}
		out[field] = v
	}
	return out
}

// ErrNoColumns is returned when the fixed-column path is used on a config that
// has no columns: mapping. Rust panics here via .expect() (transform.rs:162);
// returning an error is the Go-idiomatic equivalent and is unreachable in
// practice, since only the non-2016 path calls this.
var ErrNoColumns = errors.New("transform: config has no columns mapping")

// TransformRow extracts one row into a ParsedRow using fixed column indices,
// the 2017-family path. 2016 uses runtime format detection instead.
func TransformRow(raw []reader.Cell, cfg *config.DatasetConfig) (*ParsedRow, error) {
	cols := cfg.Columns
	if cols == nil {
		return nil, ErrNoColumns
	}

	// Trimmed accessor, mirroring the closure at transform.rs:163-167. Out-of-range
	// indices yield "" rather than an error, matching unwrap_or_default().
	get := func(idx int) string {
		if idx < 0 || idx >= len(raw) {
			return ""
		}
		return strings.TrimSpace(raw[idx].Str)
	}

	hoTen := get(cols.HoTen)
	ngaySinh := get(cols.NgaySinh)
	soBaoDanh := get(cols.SoBaoDanh)

	// diem_thi is read WITHOUT trimming (transform.rs:172-175), unlike the three
	// fields above. Harmless because the score patterns are unanchored, but it is
	// the shipped behaviour — do not "tidy" it.
	diemThi := ""
	if cols.DiemThi >= 0 && cols.DiemThi < len(raw) {
		diemThi = raw[cols.DiemThi].Str
	}

	var ngaySinhOpt *string
	if ngaySinh != "" {
		ngaySinhOpt = &ngaySinh
	}

	return &ParsedRow{
		SoBaoDanh:  soBaoDanh,
		HoTen:      hoTen,
		HoTenAscii: ToAscii(hoTen),
		NgaySinh:   ngaySinhOpt,
		TenCumThi:  nil,
		GioiTinh:   nil,
		Scores:     ParseScores(diemThi),
	}, nil
}
